core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DistinctAggregationStrategyChooser.java

```diff
@@ -0,0 +1,105 @@
+
+    private DistinctAggregationStrategyChooser(TaskCountEstimator taskCountEstimator)
+    {
+        this.taskCountEstimator = requireNonNull(taskCountEstimator, "taskCountEstimator is null");
+    }
+
+    public static DistinctAggregationStrategyChooser createDistinctAggregationStrategyChooser(TaskCountEstimator taskCountEstimator)
+    {
+        return new DistinctAggregationStrategyChooser(taskCountEstimator);
+    }
+
+    public boolean shouldAddMarkDistinct(AggregationNode aggregationNode, Rule.Context context)
+    {
+        return !canParallelizeSingleStepDistinctAggregation(aggregationNode, context, MARK_DISTINCT_MAX_OUTPUT_ROW_COUNT_MULTIPLIER);
+    }
+
+    public boolean shouldUsePreAggregate(AggregationNode aggregationNode, Rule.Context context)
+    {
+        if (canParallelizeSingleStepDistinctAggregation(aggregationNode, context, PRE_AGGREGATE_MAX_OUTPUT_ROW_COUNT_MULTIPLIER)) {
+            return false;
+        }
+
+        // mark-distinct is better than pre-aggregate if the number of group-by keys is bigger than 2
+        // because group-by keys are added to every grouping set and this makes partial aggregation behaves badly
+        return aggregationNode.getGroupingKeys().size() <= 2;
+    }
+
+    private boolean canParallelizeSingleStepDistinctAggregation(AggregationNode aggregationNode, Rule.Context context, int maxOutputRowCountMultiplier)
+    {
+        if (aggregationNode.getGroupingKeys().isEmpty()) {
+            // global distinct aggregation is computed using a single thread. MarkDistinct will help parallelize the execution.
+            return false;
+        }
+
+        double numberOfDistinctValues = getMinDistinctValueCountEstimate(aggregationNode, context);
+        if (isNaN(numberOfDistinctValues)) {
+            // if the estimate is unknown, use MarkDistinct to avoid query failure
+            return false;
+        }
+
+        int maxNumberOfConcurrentThreadsForAggregation = getMaxNumberOfConcurrentThreadsForAggregation(context.getSession());
+        if (numberOfDistinctValues <= maxOutputRowCountMultiplier * maxNumberOfConcurrentThreadsForAggregation) {
+            // small numberOfDistinctValues reduces the distinct aggregation parallelism, also because the partitioning may be skewed.
+            // This makes query to underutilize the cluster CPU but also to possibly concentrate memory on few nodes.
+            // MarkDistinct should increase the parallelism at a cost of CPU.
+            return false;
+        }
+
+        // can parallelize single-step, and single-step distinct is more efficient than alternatives
+        return true;
+    }
+
+    private int getMaxNumberOfConcurrentThreadsForAggregation(Session session)
+    {
+        return taskCountEstimator.estimateHashedTaskCount(session) * getTaskConcurrency(session);
+    }
+
+    private double getMinDistinctValueCountEstimate(AggregationNode aggregationNode, Rule.Context context)
```

Comment:
Methods in this class only care about the session and stats provider, so pass them directly. In particular, this class should not have to be aware about the rule evaluation framework (i.e., Rule.Context)